typed-protocols/src/Network/TypedProtocol/Core.hs

```diff
@@ -0,0 +1,355 @@
+-- \"from\" and \"to\" state as type parameters. This of course determines in
+-- which protocol states each message can appear.
+--
+-- In the \"ping\/pong\" protocol example, the messages are of course ping and
+-- pong, which transition between the two main states. There is also a done
+-- message that moves the system into a terminal state.
+--
+-- > instance Protocol PingPong where
+-- >   data Message PingPong from to where
+-- >     MsgPing :: Message PingPong StIdle StBusy
+-- >     MsgPong :: Message PingPong StBusy StIdle
+-- >     MsgDone :: Message PingPong StIdle StDone
+--
+-- This says that in the idle state a ping message takes us to the busy state,
+-- while a pong message takes us back to idle. Also in the idle state a done
+-- message takes us to the done state.
+--
+-- It is not required that protocols have any terminal states or corresponding
+-- transitions, as in this example, but it is often useful and it aids testing
+-- to have protocols that terminate cleanly as it allows them to return a
+-- result.
+--
+-- As described above, this style of protocol gives agency to only one peer at
+-- once. That is, in each protocol state, one peer has agency (the ability to
+-- send) and the other does not (it must only receive). The three associated
+-- data families ('ClientHasAgency', 'ServerHasAgency' and 'NobodyHasAgency')
+-- define which peer has agency for each state.
+--
+-- In the \"ping\/pong\" protocol example, the idle state is the one in which
+-- the client can send a message, and the busy state is the one in which the
+-- server must respond. Finally in the done state, neither peer can send any
+-- further messages. This arrangement is defined as so:
+--
+-- >    -- still within the instance Protocol PingPong
+-- >    data ClientHasAgency st where
+-- >      TokIdle :: ClientHasAgency StIdle
+-- >
+-- >    data ServerHasAgency st where
+-- >      TokBusy :: ServerHasAgency StBusy
+-- >
+-- >    data NobodyHasAgency st where
+-- >      TokDone :: NobodyHasAgency StDone
+--
+-- In this simple protocol there is exactly one state in each category, but in
+-- general for non-trivial protocols there may be several protocol states in
+-- each category.
+--
+
+-- $tests
+-- The mechanism for labelling each protocol state with the agency does not
+-- automatically prevent mislabelling, ie giving conflicting labels to a
+-- single state. It does in fact prevent forgetting to label states in the
+-- sense that it would not be possible to write protocol peers that make
+-- progress having entered these unlabelled states.
+--
+-- This partition property is however crucial for the framework's guarantees.
+-- The "Network.TypedProtocol.Proofs" module provides a way to guarantee for
+-- each protocol that this property is not violated. It also provides utilities
+-- helpful for testing protocols.
+
+
+-- | The protocol type class bundles up all the requirements for a typed
+-- protocol.
```

Comment:
I tried doing the above, with a type class that is indexed only on a protocol type tag. Sadly it did not work out due to a ghc issue where it doesn't do the dependency analysis within a recursive binding group correctly. A workaround is to force the ordering between the bindings but this ends up very ugly.

This could be changed so that the state kind is a type family, allowing sharing between protocols, once this ghc issue is resolved.

In the meantime it is a problem of principle but not a currently a practical problem for our use case to define the states for each protocol separately.

Reply:
For the record https://ghc.haskell.org/trac/ghc/ticket/12088

Reply:
What is the benefit of using a typeclass though? Is it just to reduce the number of parameters on a `Peer`?

Reply:
The use of data families reduces the number of parameters (particularly the confusing ones). The use of a type class with associated data types (rather than stand alone data families) makes it clearer what one needs to do to implement a protocol by bundling the related things together. That aspect is just organisational. e.g. we get clearer documentation, you can see in the haddock docs or ghci what all the `Protocol` instances are. It pulls the constituent parts of the interface together into one unit.

Reply:
> The use of data families reduces the number of parameters (particularly the confusing ones).

I really object to this notion that they are confusing. Is there any data to support that? It's 3 parameters: local partition, remote partition, terminal partition. We can document that, it's quite simple. Honestly, introducing type families is likely to confuse more users.

> we get clearer documentation, you can see in the haddock docs or ghci what all the Protocol

Should protocols be global, as typeclasses are? I see no benefit in this and in fact I think it's quite bizarre.

> It pulls the constituent parts of the interface together into one unit.

Could be done with modules and type synonyms.

```Haskell
import GenericRequestResponse (Client, Server, Terminal, TrReqRes)

type MyClientApp = Peer Client Server Terminal (TrReqRes Int Int)
type MyServerApp = Peer Server Client Terminal (TrReqRes Int Int)
```